src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+                                "already contains a symbol that clashes with "
+                                "one of those ('{1}') that must be imported "
+                                "from the PSyclone PSyData module.".
+                                format(routine_name, symbol))
+                    # Check for the name of the PSyData module itself
+                    if text == self.fortran_module:
+                        raise NotImplementedError(
+                            "Cannot add PSyData calls to '{0}' because it "
+                            "already contains a symbol that clashes with the "
+                            "name of the PSyclone PSyData module "
+                            "('{1}')". format(routine_name,
+                                              self.fortran_module))
+                    # Check for the names of PSyData variables
+                    if text.startswith(self.psy_data_var):
+                        raise NotImplementedError(
+                            "Cannot add PSyData calls to '{0}' because it "
+                            "already contains symbols that potentially clash "
+                            "with the variables we will insert for each "
+                            "PSyData region ('{1}*').".
+                            format(routine_name, self.psy_data_var))
+        # Flag that we have now checked for name clashes so that if there's
+        # more than one PSyData node we don't fall over on the symbols
+        # we've previous inserted.
+        self.root.psy_data_name_clashes_checked = True
+
+        # Create a name for this region by finding where this PSyDataNode
+        # is in the list of PSyDataNodes in this Invoke.
+        sched = self.root
+        pnodes = sched.walk(PSyDataNode)
+        region_idx = pnodes.index(self)
+        if self._region_name:
+            region_name = self._region_name
+        else:
+            region_name = "r{0}".format(region_idx)
+        var_name = "psy_data{0}".format(region_idx)
+
+        # Create a variable for this PSyData region
+        reader = FortranStringReader(
+            "type(PSyDataType), save :: {0}".format(var_name))
+        # Tell the reader that the source is free format
+        reader.set_format(FortranFormat(True, False))
+        decln = Fortran2003.Type_Declaration_Stmt(reader)
+        spec_part.content.append(decln)
+
+        # Find the parent in the parse tree - first get a pointer to the
+        # AST for the content of this region.
+        content_ast = self.psy_data_body.children[0].ast
+        # Now store the parent of this region
+        # pylint: disable=protected-access
+        fp_parent = content_ast.parent
+        # pylint: enable=protected-access
+        # Find the location of the AST of our first child node in the
+        # list of child nodes of our parent in the fparser parse tree.
+        ast_start_index = object_index(fp_parent.content,
```

Comment:
To a certain extent this code duplicates that of `_add_region` in psyGen.py, albeit that that code expects to add directives rather than the more complex code added here. Of course, once #435 is done (use the PSyIR backend for the NEMO API) we will get rid of this code altogether. The question is, should we put effort into refactoring now or live with the code duplication? I've lost track of how far off #435 is. Maybe @rupertford has a clearer view?

Reply:
I am strongly in favour of not putting effort into code that we know will become obsolete in the (not too distant) future.